### ISSUE TITLE
what about the area in target(t) that do not have a correspondence pixel in the source(t’)?

### ISSUE THREAD

**Issue by biggiantpigeon** [nianticlabs/monodepth2#357]
I have a question about reprojection loss calculation.
In my understanding, In function generate_images_pred, the code calculate the depth of target image, and then find the corresponding pixel in source image(by bilinear interpolation). But if the camera is not static, the warped image must always have some area which do not have a corresponding pixel from the source, and when the padding_mode in F.grid_sample is 'zeros', that area's pixel value is set to zero in I_t’->t._
I don't see any mask dealing with these area, and in my understanding, they also take part in the SSIM loss calculation. Isn't this unreasonable, or do I miss something?

**Comment by biggiantpigeon** [comment]
![image](https://user-images.githubusercontent.com/80443901/121282350-eb092380-c90b-11eb-96ec-e821816748ae.png)

since no one is asking in days, I wonder if I need to state the question more precisely. Please see the image above.

the left image is the source, and the right image is warped source with predict depth and pose. photometric reprojection error is calculated with the target image(middle) and the warped(right). what about the zero(black) area in the warped image? They also take part in the loss, right? @mdfirman

**Comment by mdfirman** [comment]
Hi – thanks for your interest here.

I understand the question you are asking, and thanks for the image to make this clear!

You are right that the loss is applied to all pixels in the image. We don't do any form of masking to remove the loss from the pixels in the 'black area' you have shown in your image above.

However, this isn't a problem. The loss will try to make the black region of $I_{t' -> t}$ look like the corresponding region of $I_t$. But because this black region is in fact outside the bounds of $I_{t'}$, no gradient can be backpropogated through these pixels. So the loss here doesn't have any effect on the gradients going through the networks. So the loss computed from these pixels effectively becomes a 'constant' in the loss for this iteration, and the network can only learn to improve depth and pose based on the valid (i.e. non-black) pixels

(In practice we don't use zero padding for out of range pixels – we use `padding_mode='border'` in [grid sample](https://pytorch.org/docs/stable/nn.functional.html#grid-sample), but the concept is the same)

Does this help?

**Comment by biggiantpigeon** [comment]
Thanks, I understand it now.
